endpoints/openrtb2/auction.go

```diff
@@ -279,6 +279,9 @@ func (deps *endpointDeps) Auction(w http.ResponseWriter, r *http.Request, _ http
 		glog.Errorf("Error setting seat non-bid: %v", err)
 	}
 	labels, ao = sendAuctionResponse(w, hookExecutor, response, req.BidRequest, account, labels, ao)
+	if len(ao.Errors) == 0 && !auctionRequest.BidderResponseStartTime.IsZero() {
+		deps.metricsEngine.RecordOverheadTime(metrics.MakeAuctionResponse, time.Since(auctionRequest.BidderResponseStartTime))
+	}
```

Comment:
From a previous conversation with @hhhjort , my understanding is that we want the `metrics.MakeAuctionResponse` metric to record the time it takes to process the auction response if PBS made at least one roundtrip to an adapter's bid server regardless of the number of errors and/or warnings we get in the end. Therefore, as long as we were able to send a request and receive a response from a bid server, we want to have an statistic of how long it takes for PBS to build the response, even if that response ends up being an error, because handling and returning an error is part of PBS core. Under this logic, I believe we should:
1) Remove the first part of the check. We know `auctionRequest.BidderResponseStartTime` is set if at least one `httpInfo.err` was `nil` in line 244 of `exchange/bidder.go`:
```
286 -   if len(ao.Errors) == 0 && !auctionRequest.BidderResponseStartTime.IsZero() {
    +   if !auctionRequest.BidderResponseStartTime.IsZero() {
287         deps.metricsEngine.RecordOverheadTime(metrics.MakeAuctionResponse, time.Since(auctionRequest.BidderResponseStartTime))
288     }
endpoints/openrtb2/auction.go
```
2) Make sure this chunk of code is run regardless where we exit the handler function. I believe the `defer` statement might be useful:
```
254       auctionResponse, err := deps.ex.HoldAuction(ctx, auctionRequest, nil)
    +     defer func() {
    +         if !auctionRequest.BidderResponseStartTime.IsZero() {
    +             deps.metricsEngine.RecordOverheadTime(metrics.MakeAuctionResponse, time.Since(auctionRequest.BidderResponseStartTime))
    +         }
    +     }()
255       ao.RequestWrapper = req
256       ao.Account = account
257       var response *openrtb2.BidResponse
258       if auctionResponse != nil {
259           response = auctionResponse.BidResponse
260       }
261       ao.Response = response
262       ao.SeatNonBid = auctionResponse.GetSeatNonBid()
263       rejectErr, isRejectErr := hookexecution.CastRejectErr(err)
264       if err != nil && !isRejectErr {
265           if errortypes.ReadCode(err) == errortypes.BadInputErrorCode {
266               writeError([]error{err}, w, &labels)
267               return
268           }
269           labels.RequestStatus = metrics.RequestStatusErr
270           w.WriteHeader(http.StatusInternalServerError)
271           fmt.Fprintf(w, "Critical error while running the auction: %v", err)
272           glog.Errorf("/openrtb2/auction Critical error: %v", err)
273           ao.Status = http.StatusInternalServerError
274           ao.Errors = append(ao.Errors, err)
275           return
276       } else if isRejectErr {
277           labels, ao = rejectAuctionRequest(*rejectErr, w, hookExecutor, req.BidRequest, account, labels, ao)
278           return
279       }
280  
281       err = setSeatNonBidRaw(req, auctionResponse)
282       if err != nil {
283           glog.Errorf("Error setting seat non-bid: %v", err)
284       }
285       labels, ao = sendAuctionResponse(w, hookExecutor, response, req.BidRequest, account, labels, ao)
286 -     if len(ao.Errors) == 0 && !auctionRequest.BidderResponseStartTime.IsZero() {
287 -         deps.metricsEngine.RecordOverheadTime(metrics.MakeAuctionResponse, time.Since(auctionRequest.BidderResponseStartTime))
288 -     }
289   }
endpoints/openrtb2/auction.go
```
Do you agree with logging the metric even if we return errors? If so, this change should also be implemented in `endpoints/openrtb2/amp_auction.go` and `endpoints/openrtb2/video_auction.go`. Let me know your thoughts.

Reply:
Thanks @guscarreon and @hhhjort 

https://github.com/prebid/prebid-server/pull/2841/commits/7795eec3a4743e453a83c8d7e7f38b56aaa81946 makes the change